### PR TITLE
Endpoint resolution engine bindings

### PR DESCRIPTION
**Pull request by DmitriyMusatkin**
*Issue #, if available:*

*Description of changes:*

Endpoint resolution engine bindings

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: include/aws/crt/endpoints/RuleEngine.h [hunk 1]
```diff
@@ -0,0 +1,157 @@
...
+             *   endpoint could not be resolved.
+             */
+            class AWS_CRT_CPP_API ResolutionOutcome final
+            {
+              public:
+                ~ResolutionOutcome();
+
+                /* TODO: move/copy semantics */
+                ResolutionOutcome(const ResolutionOutcome &) = delete;
+                ResolutionOutcome &operator=(const ResolutionOutcome &) = delete;
```
Comment (SergeyRyabinin) [inline include/aws/crt/endpoints/RuleEngine.h:86]: very minor nit
why & next to operator not the type?

File: source/endpoints/RuleEngine.cpp [hunk 2]
```diff
@@ -0,0 +1,177 @@
...
+
+            Optional<StringView> ResolutionOutcome::GetUrl() const noexcept
+            {
+                ByteCursor url;
+                if (aws_endpoints_resolved_endpoint_get_url(m_resolvedEndpoint, &url))
+                {
+                    return Optional<StringView>();
+                }
+
+                return Optional<StringView>(ByteCursorToStringView(url));
```
Comment (SergeyRyabinin) [inline source/endpoints/RuleEngine.cpp:75]: both c-tors with arguments of Crt::Optional
```
Optional(const T &val)
Optional(T &&val)
```
have a call to `new`
and `new` can throw at least std::bad_alloc

File: source/endpoints/RuleEngine.cpp [hunk 3]
```diff
@@ -0,0 +1,177 @@
...
+            }
+
+            inline StringView CrtStringToStringView(const aws_string *s)
+            {
+                ByteCursor key = aws_byte_cursor_from_string(s);
+                return ByteCursorToStringView(key);
+            }
+
+            Optional<UnorderedMap<StringView, Vector<StringView>>> ResolutionOutcome::GetHeaders() const noexcept
```
Comment (SergeyRyabinin) [inline source/endpoints/RuleEngine.cpp:84]: this function is declared as `noexcept`
while it obviously allocate memory using `new` in both Optional and UnorderedMap and Vector.

File: include/aws/crt/endpoints/RuleEngine.h (hunk 1 above)
Comment (DmitriyMusatkin) [reply include/aws/crt/endpoints/RuleEngine.h:86]: dont know the history of this, but this is the format used across the package, so just staying consistent

File: source/endpoints/RuleEngine.cpp [hunk 4]
```diff
@@ -0,0 +1,177 @@
...
+                    return Optional<UnorderedMap<StringView, Vector<StringView>>>();
+                }
+
+                UnorderedMap<StringView, Vector<StringView>> headers;
+                for (struct aws_hash_iter iter = aws_hash_iter_begin(resolved_headers); !aws_hash_iter_done(&iter);
+                     aws_hash_iter_next(&iter))
+                {
+                    ByteCursor key = aws_byte_cursor_from_string((const aws_string *)iter.element.key);
+                    const aws_array_list *array = (const aws_array_list *)iter.element.value;
+                    headers.insert(
```
Comment (SergeyRyabinin) [inline source/endpoints/RuleEngine.cpp]: although `insert` may take the most efficient copy-elision overload,
have you tried/considered `emplace` to be sure that copy of a temporary object does not happen?

File: source/endpoints/RuleEngine.cpp (hunk 2 above)
Comment (DmitriyMusatkin) [reply source/endpoints/RuleEngine.cpp:75]: hmm, good point. i see we are doing noexcept in other cases where we create optional (ex. http classes). will remove noexcept

File: source/endpoints/RuleEngine.cpp (hunk 4 above)
Comment (DmitriyMusatkin) [reply source/endpoints/RuleEngine.cpp]: yeah, emplace would be better here.

**Review by SergeyRyabinin — APPROVED** [review]

**Review by TingDaoK — APPROVED** [review]
LGTM

**Pull request merged**